packages/graphback-runtime/src/service/ServiceCreator.ts

```diff
@@ -0,0 +1,10 @@
+import { GraphbackDataProvider } from '../data/GraphbackDataProvider';
+import { PubSubConfig } from '../GraphbackPubSubModel';
+import { GraphbackCRUDService } from './GraphbackCRUDService';
+
+/**
+ * Abstract class for creating a new CRUD Service
+ */
+export abstract class ServiceCreator {
+  public abstract createService(modelName: string, dataProvider: GraphbackDataProvider, pubSubConfig: PubSubConfig): GraphbackCRUDService;
```

Comment:
Help wanted! Actually I do not see how this would work. This `createService` method assumes a very strict signature.

Am I implementing this wrong @wtrocki ?

Reply:
We have agreed on the creator pattern, however this is a bit ambiguous as there seems to be multiple creator pattern types (factory, builder being two).

I've implemented factory pattern, because I believe we discussed this. However I can't make it generic enough to work for all cases.

Reply:
I think Builder pattern makes more sense, and is more flexible.

Reply:
I did not clearly explain that so happy to jump to some call to avoid misunderstanding. 

WHat I mean that both `ServiceCreator` and `createService` abstraction is redundant. 
You can see this as callback and the only granted values that will be generic in the callback would be `model`?

Reply:
Think I get you now. Will get back to this tomorrow!

Reply:
+1 for a function.

Reply:
>  the only granted values that will be generic in the callback would be model?

I don't think this is guaranteed to always be the case.

Look at KeycloakCrudService. The constructor signature is different to the standard CRUD service and it has other model config specific to this service which we cannot account for at runtime.
 
https://github.com/aerogear/graphback/blob/0fdc6fb7babe4b70831e9140efe6bb4d2057aae5/packages/graphback-keycloak-authz/src/KeycloakCrudService.ts#L35

Reply:
See that authConfig is custom to that model. 

Graphback has to know the structure of the service when creating it:

https://github.com/aerogear/graphback/blob/57fb74e082f213cbe4c9f109ed8387b0d81a1f60/packages/graphback/src/buildGraphbackAPI.ts#L64

But this won't work for KeycloakCRUDService.

I kind of feel like it the original helpers like `createKnexCRUD...RuntimeServices` were better simply because they have the final schema and it is easy to create the services and data providers. 

Yes they don't allow overriding of the services, and it involves having to create one if you want to use a custom services and/or provider.

I will need to think about this.

Reply:
Actually I might be wrong about that.

Reply:
>  Look at KeycloakCrudService. The constructor signature is different to the standard CRUD service and it has other model config specific to this service which we cannot account for at runtime.

I think we failing to understand each other here. Call badly needed.